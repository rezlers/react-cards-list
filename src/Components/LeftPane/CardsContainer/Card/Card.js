import React from 'react';
import styles from './Card.module.css'

function Card(props) {
    const {deleteCard, value} = props;
    return (
        <div className={`${styles['custom-card']}`}>
            <i className={`fa fa-close ${styles['delete-icon']}`} onClick={deleteCard}></i>
            {value}
        </div>
    );
}

export default Card
