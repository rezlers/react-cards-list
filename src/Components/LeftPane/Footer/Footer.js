import React from 'react';

function Footer() {
    return (
        <div className='footer-copyright text-center py-3 border border-secondary'>
            <p>Footer</p>
        </div>
    );
}

export default Footer
