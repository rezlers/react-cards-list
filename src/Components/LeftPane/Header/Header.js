import React from 'react';

function Header(props) {
    const {sortCards, addCard} = props;

    return (
        <div className="d-flex justify-content-start align-items-center border border-secondary">
            <button className='btn btn-secondary mx-1 my-2' onClick={sortCards}>Sort cards</button>
            <button className='btn btn-secondary mx-1 my-2' onClick={addCard}>Add card</button>
        </div>
    );
}

export default Header
