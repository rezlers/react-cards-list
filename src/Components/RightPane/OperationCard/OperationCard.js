import React from 'react';

function OperationCard(props) {
    const {msg} = props;

    return (
        <div className='card'>
            <p>{msg}</p>
        </div>
    );
}

export default OperationCard
