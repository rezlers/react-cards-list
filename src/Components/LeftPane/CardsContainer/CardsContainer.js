import React from 'react';
import Card from "./Card/Card";
import styles from './CardsContainer.module.css'

function CardsContainer(props) {
    const {cardsList, deleteCard} = props;

    return (
        <div className={`${styles['cards-container']} overflow-auto border border-secondary`}>
            {
                cardsList.map((item, index) => {
                    return (
                        <>
                            <Card
                                value={item}
                                deleteCard={deleteCard(index)}
                                key={index}
                            />
                        </>
                    )
                })
            }
        </div>
    );
}

export default CardsContainer
