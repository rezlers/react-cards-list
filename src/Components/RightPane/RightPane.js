import React from 'react';
import OperationCard from "./OperationCard/OperationCard";
import styles from './RightPane.module.css'

class RightPane extends React.Component {
    render() {
        return (
            <div className={`${styles['right-pane']} overflow-auto `}>
                {
                    this.props.operations.map((item, index) => {
                        return (
                            <>
                                <OperationCard
                                    key={index}
                                    msg={item}
                                />
                            </>
                        )
                    })
                }
            </div>
        );
    }
}

export default RightPane
