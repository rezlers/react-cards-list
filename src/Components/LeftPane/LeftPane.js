import React from 'react';
import Footer from "./Footer/Footer";
import Header from "./Header/Header";
import CardsContainer from "./CardsContainer/CardsContainer";
import styles from './LeftPane.module.css'

class LeftPane extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            cardsList: []
        };
        this.addCard = this.addCard.bind(this);
        this.deleteCard = this.deleteCard.bind(this);
        this.sortCards = this.sortCards.bind(this)
    }

    randInt(min, max) {
        min = Math.ceil(min);
        max = Math.floor(max);
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    addCard() {
        const newCardsList = [this.randInt(0,100)].concat(this.state.cardsList);
        this.setState({cardsList: newCardsList});
        this.props.onAdd(newCardsList.length - 1)
    };

    sortCards() {
        this.setState({cardsList: this.state.cardsList.concat().sort(function (a, b) {
            return b - a;
        })});
        this.props.onSort();
    };

    deleteCard(cardIndex) {
        return () => {
            const newCardsList = this.state.cardsList.filter((item, index) => {
                    return index !== cardIndex;
                });
            this.setState({cardsList: newCardsList});
            this.props.onDelete(cardIndex)
    }};

    render() {
        return (
            <div className={styles['left-pane']}>
                <Header
                    addCard={this.addCard}
                    sortCards={this.sortCards}
                />
                <CardsContainer
                    cardsList={this.state.cardsList}
                    deleteCard={this.deleteCard}
                />
                <Footer/>
            </div>
        );
    }
}

export default LeftPane
