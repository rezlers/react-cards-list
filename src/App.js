import React from 'react';
import ReactDOM from 'react-dom';
import RightPane from "./Components/RightPane/RightPane";
import LeftPane from "./Components/LeftPane/LeftPane";
import styles from './App.module.css'

class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = {operations: []}
    }

    render() {
        return (
            <div className={`${styles['app-container']} container-fluid`}>
                <div className='row'>
                    <div className='col-lg-9 col-md-9 col-sm-9 col-xs-12'>
                        <LeftPane
                            onDelete={(index) => this.setState({operations: [...this.state.operations, `Delete card: ${index}`]})}
                            onAdd={(index) => this.setState({operations: [...this.state.operations, `Add card: ${index}`]})}
                            onSort={() => this.setState({operations: [...this.state.operations, `Sort cards`]})}
                        />
                    </div>
                    <div className='col-lg-3 col-md-3 col-sm-3 d-none d-sm-block border border-secondary'>
                        <RightPane
                            operations={this.state.operations}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

export default App;

if (document.getElementById('root')) {
    ReactDOM.render(<App />, document.getElementById('root'));
}

