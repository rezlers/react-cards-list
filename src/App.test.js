import React from 'react';
import { shallow, configure, mount } from 'enzyme'
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import LeftPane from "./Components/LeftPane/LeftPane";
import RightPane from "./Components/RightPane/RightPane";
import Card from "./Components/LeftPane/CardsContainer/Card/Card";
import CardsContainer from "./Components/LeftPane/CardsContainer/CardsContainer";
import Header from "./Components/LeftPane/Header/Header";
import Footer from "./Components/LeftPane/Footer/Footer";
import OperationCard from "./Components/RightPane/OperationCard/OperationCard";

configure({ adapter: new Adapter() });

describe('Testing adding card process', () => {
  test('The function addCard() is working correctly', () => {
    const component = mount(<LeftPane
        onDelete = {() => {}}
        onAdd = {() => {}}
        onSort = {() => {}}
    />);
    const instance = component.instance();
    instance.addCard();
    expect(instance.state.cardsList[0]).toBeTruthy();
  });

  test('Card component renders correctly', () => {
      const component = mount(<LeftPane
          onDelete = {() => {}}
          onAdd = {() => {}}
          onSort = {() => {}}
      />);
      const instance = component.instance();
      instance.setState({
          cardsList: [1]
      });
      const tree = renderer
          .create(<Card
              value={instance.state.cardsList[0]}
              deleteCard={instance.deleteCard(0)}
          />)
          .toJSON();
      expect(tree).toMatchSnapshot();
  });

});

describe('Testing deleting card process', () => {
    test('The function deleteCard() is working correctly', () => {
        const component = mount(<LeftPane
            onDelete = {() => {}}
            onAdd = {() => {}}
            onSort = {() => {}}
        />);
        const instance = component.instance();
        instance.addCard();
        expect(instance.state.cardsList[0]).toBeTruthy();
        instance.deleteCard(0)();
        expect(instance.state.cardsList[0]).toBeFalsy();

        instance.addCard();
        instance.addCard();
        instance.addCard();
        instance.deleteCard(1)();
        expect(instance.state.cardsList.length).toBe(2);
    });

});

describe('Testing sorting cardsList process', () => {
    test('The function sortCards() is working correctly', () => {
        const CARDS_NUM = 5;

        const component = mount(<LeftPane
            onDelete = {() => {}}
            onAdd = {() => {}}
            onSort = {() => {}}
        />);
        const instance = component.instance();
        do {
            for(let i = 0; i < CARDS_NUM; i++) {
                instance.addCard();
            }
        } while (instance.state.cardsList === instance.state.cardsList.concat().sort((a, b) => {
            return b - a
        }));
        instance.sortCards();
        expect(instance.state.cardsList).toMatchObject(instance.state.cardsList.concat().sort((a, b) => {
            return b - a
        }));
    });

});

describe('Testing adding value to operations array when addCard/deleteCard/sortCards methods is called', () => {
    test('The function addCard() is working correctly', () => {
        const appComponent = mount(<App />);
        const appInstance = appComponent.instance();

        appComponent.find(LeftPane).instance().addCard();
        expect(appInstance.state.operations).toMatchObject(["Add card: 0"]);

        appComponent.find(LeftPane).instance().deleteCard(0)();
        expect(appInstance.state.operations).toMatchObject(["Add card: 0", "Delete card: 0"]);

        appComponent.find(LeftPane).instance().setState({
            cardsList: [1, 2, 10, 99, 6, 17, 25, 1, 5]
        });

        appComponent.find(LeftPane).instance().sortCards();
        expect(appInstance.state.operations).toMatchObject(["Add card: 0", "Delete card: 0", "Sort cards"]);
    });

});

describe('CardsContainer renders correctly on adding/deleting/sorting values in CardsList', () => {
    test('CardsContainer renders correctly after adding a new node', () => {
        const component = shallow(<LeftPane
            onDelete = {() => {}}
            onAdd = {() => {}}
            onSort = {() => {}}
        />);
        const instance = component.instance();
        instance.setState({
            cardsList: [1]
        });
        const tree1 = renderer
            .create(<CardsContainer
                cardsList={instance.state.cardsList}
                deleteCard={instance.deleteCard}
            />)
            .toJSON();
        expect(tree1).toMatchSnapshot();
        instance.setState({
            cardsList: [1, 2]
        });
        const tree2 = renderer
            .create(<CardsContainer
                cardsList={instance.state.cardsList}
                deleteCard={instance.deleteCard}
            />)
            .toJSON();
        expect(tree2).toMatchSnapshot()
    });

    test('CardsContainer renders correctly after deleting a node', () => {
        const component = shallow(<LeftPane
            onDelete = {() => {}}
            onAdd = {() => {}}
            onSort = {() => {}}
        />);
        const instance = component.instance();
        instance.setState({
            cardsList: [1, 2]
        });
        instance.deleteCard(1)();
        const tree = renderer
            .create(<CardsContainer
                cardsList={instance.state.cardsList}
                deleteCard={instance.deleteCard}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('CardsContainer renders correctly before and after sorting', () => {
        const component = shallow(<LeftPane
            onDelete = {() => {}}
            onAdd = {() => {}}
            onSort = {() => {}}
        />);
        const instance = component.instance();
        instance.setState({
            cardsList: [1, 2, 10, 99, 6, 17, 25, 1, 5]
        });
        const tree1 = renderer
            .create(<CardsContainer
                cardsList={instance.state.cardsList}
                deleteCard={instance.deleteCard}
            />)
            .toJSON();
        expect(tree1).toMatchSnapshot();
        instance.sortCards();
        const tree2 = renderer
            .create(<CardsContainer
                cardsList={instance.state.cardsList}
                deleteCard={instance.deleteCard}
            />)
            .toJSON();
        expect(tree2).toMatchSnapshot();
    });
});

describe('Testing if single components renders correctly', () => {
    test('Testing Header', () => {
        const tree = renderer
            .create(<Header
                addCard={() => {}}
                sortCards={() => {}}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Testing Footer', () => {
        const tree = renderer
            .create(<Footer
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Testing RightPane', () => {
        const tree = renderer
            .create(<RightPane
                operations = {["Add card: 0", "Delete card: 0", "Sort cards"]}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('Testing OperationCard', () => {
        const tree = renderer
            .create(<OperationCard
                msg = "Sort cards"
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

