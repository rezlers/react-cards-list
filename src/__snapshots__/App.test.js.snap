// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer renders correctly on adding/deleting/sorting values in CardsList CardsContainer renders correctly after adding a new node 1`] = `
<div
  className="cards-container overflow-auto border border-secondary"
>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
</div>
`;

exports[`CardsContainer renders correctly on adding/deleting/sorting values in CardsList CardsContainer renders correctly after adding a new node 2`] = `
<div
  className="cards-container overflow-auto border border-secondary"
>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    2
  </div>
</div>
`;

exports[`CardsContainer renders correctly on adding/deleting/sorting values in CardsList CardsContainer renders correctly after deleting a node 1`] = `
<div
  className="cards-container overflow-auto border border-secondary"
>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
</div>
`;

exports[`CardsContainer renders correctly on adding/deleting/sorting values in CardsList CardsContainer renders correctly before and after sorting 1`] = `
<div
  className="cards-container overflow-auto border border-secondary"
>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    2
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    10
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    99
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    6
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    17
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    25
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    5
  </div>
</div>
`;

exports[`CardsContainer renders correctly on adding/deleting/sorting values in CardsList CardsContainer renders correctly before and after sorting 2`] = `
<div
  className="cards-container overflow-auto border border-secondary"
>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    99
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    25
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    17
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    10
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    6
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    5
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    2
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
  <div
    className="custom-card"
  >
    <i
      className="fa fa-close delete-icon"
      onClick={[Function]}
    />
    1
  </div>
</div>
`;

exports[`Testing adding card process Card component renders correctly 1`] = `
<div
  className="custom-card"
>
  <i
    className="fa fa-close delete-icon"
    onClick={[Function]}
  />
  1
</div>
`;

exports[`Testing if single components renders correctly Testing Footer 1`] = `
<div
  className="footer-copyright text-center py-3 border border-secondary"
>
  <p>
    Footer
  </p>
</div>
`;

exports[`Testing if single components renders correctly Testing Header 1`] = `
<div
  className="d-flex justify-content-start align-items-center border border-secondary"
>
  <button
    className="btn btn-secondary mx-1 my-2"
    onClick={[Function]}
  >
    Sort cards
  </button>
  <button
    className="btn btn-secondary mx-1 my-2"
    onClick={[Function]}
  >
    Add card
  </button>
</div>
`;

exports[`Testing if single components renders correctly Testing OperationCard 1`] = `
<div
  className="card"
>
  <p>
    Sort cards
  </p>
</div>
`;

exports[`Testing if single components renders correctly Testing RightPane 1`] = `
<div
  className="right-pane overflow-auto "
>
  <div
    className="card"
  >
    <p>
      Add card: 0
    </p>
  </div>
  <div
    className="card"
  >
    <p>
      Delete card: 0
    </p>
  </div>
  <div
    className="card"
  >
    <p>
      Sort cards
    </p>
  </div>
</div>
`;
